extensions/transaction/transaction-local/src/main/java/org/eclipse/dataspaceconnector/transaction/local/LocalDataSourceRegistry.java

```diff
@@ -14,36 +14,41 @@
 package org.eclipse.dataspaceconnector.transaction.local;
 
 import org.eclipse.dataspaceconnector.spi.transaction.datasource.DataSourceRegistry;
-import org.eclipse.dataspaceconnector.spi.transaction.local.LocalTransactionContextManager;
 
 import java.util.HashMap;
 import java.util.Map;
 import java.util.Objects;
+import java.util.Set;
+import java.util.function.BiConsumer;
 import javax.sql.DataSource;
 
 /**
  * Default {@link DataSourceRegistry} implementation. This implementation is used to register both local and XA data sources so that they can be resolved by extensions.
  */
 public class LocalDataSourceRegistry implements DataSourceRegistry {
-    private LocalTransactionContextManager manager;
```

Comment:
Hey @denisneuling can you explain why these changes were made to the LocalDataSourceRegistry?

Reply:
Sorry, I'm confused. Here is what I think should happen:

1. Each configuration of the ConnectionPool implementation registers exactly one DataSource that is backed by the pool during extension initialization.
2. This pooled DataSource implementation will be wrapped in a DataSourceResource and registered with the LocalTransactionContext, which acts as a transaction manager. 
3. DataSourceResource is threadsafe. When the getConnection() method is called, the DataSourceResource will delegate to the connection pool DataSource implementation if no connection is enlisted for the thread.  
4. The connection pool DataSource will lease and return a Connection from the pool. This connection instance will be pinned to the current thread by DataSourceResource.
5. Subsequent calls to return a connection within the same transaction context will return the same leased Connection.

Is there something I am missing that would cause this not to work?

Reply:
That's how it's working except that the `CommonsConnectionPoolServiceExtension` does rely on the `DataSourceRegistry` to provide it with already existing `DataSource`s to be wrapped.

1. The `CommonsConnectionPoolServiceExtension` [iterates through the config properties to find all configured `DataSource`s](https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/pull/499/files#diff-e879c6194f1780828137ba3c349f945e819d85cbb2bbaf488c5c3ac92102ce74R68-R80) 
2. Those `DataSource`s are [retrieved from the `DataSourceRegistry`, if found wrapped into a `ConnectionPoolDataSource` and again re-registered (overridden) at the `DataSourceRegistry`](https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/pull/499/files#diff-e879c6194f1780828137ba3c349f945e819d85cbb2bbaf488c5c3ac92102ce74R52-R56)
3. All `DataSource`s known to the `LocalDataSourceRegistry` become [registered at the `LocalTransactionContext` after `LocalTransactionExtension` intialization](https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/pull/499/files#diff-f37781325ab23fa4fa2ea8cf863e4c9a7c505cb54ff440a807b2f4aff8848e1bR51-R54)

If that's still unclear I'd be happy having a call to lead you through the code.

Reply:
I think I understand what the issue is. It looks as if you are trying to wrap *existing* connections in a pool that is then registered. That's not how it should work.

Instead, a pool implementation is responsible for creating a delegate DataSource from configuration, instantiating it, and giving it to the pool. The pool will then be registered in Step 1 that I outlined above.

Everything should then work fine and the changes made to the local transaction extension can be reverted.

Reply:
Thought the only responsibility was be to provide pooling functionality without actually instantiating `Connection`s by calling the `DriverManager`. Seems I misunderstood the intention of that extension then...  Will then remove the `:extension:sql:postgresql` at the same adding the proposed functionality.